torch/dispatch/dispatcher.py

```diff
@@ -0,0 +1,95 @@
+import torch
+import torch._C as _C
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard  # type: ignore[attr-defined]
+from torch.utils._pytree import tree_flatten
+from torch.overrides import handle_torch_function, has_torch_function
+
+SUPPORTED_KEYS = {
+    DispatchKey.CPU,
+    DispatchKey.BackendSelect,
+    DispatchKey.ADInplaceOrView,
+    DispatchKey.AutogradCPU,
+    DispatchKey.Python,
+    DispatchKey.PythonTLSSnapshot,
+}
+
+"""
+This is a dispatcher (in Python)
+- You can define new operations (in Python) without schemas
+- It interfaces with the PyTorch dispatcher
+"""
+
+class PyDispatcher:
+    def __init__(self):
+        self.current_dispatching_op = None
+        self.already_dispatched_keys = None
+        self.current_key = None
+
+    def call(self, operator, args, kwargs):
+        try:
+            key = compute_dispatch_key(operator, args, kwargs, self.current_key)
+            self.record_dispatch(key, operator)
+            return dispatch(key, operator, args, kwargs)
+        finally:
+            self.reset_dispatch_record()
+
+    def redispatch(self, operator, args, kwargs):
+        # Redispatch doesn't go to the top
+        assert operator == self.currently_dispatching_op
+        key = compute_dispatch_key(operator, args, kwargs, self.current_key, self.already_dispatched_keys)
+        self.record_dispatch(key, operator)
+        return dispatch(key, operator, args, kwargs)
+
+    def reset_dispatch_record(self):
+        self.current_dispatching_op = None
+        self.already_dispatched_keys = None
+
+    def record_dispatch(self, dispatch_key, operator):
+        self.currently_dispatching_op = operator
+        self.current_key = dispatch_key
+        if self.already_dispatched_keys is None:
+            self.already_dispatched_keys = DispatchKeySet(dispatch_key)
+        else:
+            self.already_dispatched_keys = self.already_dispatched_keys | DispatchKeySet(dispatch_key)
+
+
+dispatcher_singleton = PyDispatcher()
+
+
+def compute_dispatch_key(operator, args, kwargs, current_key, additional_exclude=None):
+    if current_key is not None and operator.entrance_rules[current_key]:
+        return current_key
+
+    tensors = get_tensors(args, kwargs)
+    dispatch_key = key_extractor(tensors, additional_exclude)
+    return dispatch_key
+
+
+def dispatch(dispatch_key, operator, args, kwargs):
+    if dispatch_key not in SUPPORTED_KEYS:
+        raise RuntimeError(f'NYI: {dispatch_key} {SUPPORTED_KEYS}')
+    assert dispatch_key in operator.table
+    kernel = operator.table[dispatch_key]
+    return kernel(*args, **kwargs)
+
+
+def key_extractor(tensors, additional_exclude=None):
```

Comment:
Cross-ref this with the relevant C++ implementation

Reply:
So the difference between this and the c++ impl of key extractor is somewhat present, but does not feel super significant. Highlighting them below

1. sets `nonFallthroughKeys_` at ctor
2. creates `nonFallthroughKeysPerBackend_` (we do not support this, yet, at all) 
3. computes a bitset for the given function schema
4. returns a `getDispatchKeySetUnboxed` which is analogous to our `key_set` here - both are a keyset, having gone through inclusion and then exclusion 

The biggest difference is here we call `_dispatch_keys` per tensor, in the args as opposed to `ks = ks | ivalue.unsafeToTensorImpl()->key_set();` which uses `torch::jit::peek` to get the Ivalues out.

This solution is better, because we can just unpack tensors in py land and extract keys as needed - we are effectively doing the same thing but without the ivalue and c++ indirection and management that comes with dealing with Ivalue (ex: see the `isTensorList` and `isList` logic for peeking). More importantly this seems philosophically identical with the primary distinction of not really supporting fallthroughs per backend yet.

Reply:
Sorry, when I say "crossref", I mean, put a code pointer to the analogous C++ dispatcher code (and vice versa), so people know what other sites to update if they change the logic.